### PR TITLE
All subscribers report by subscription voucher types

### PR DESCRIPTION
**Pull request by AlanHe-Xiaoyu**
For Reports' All Subscribers, change from season tags to voucher-type tags

### CODE REVIEW

**Comment by Hexhu** [comment]
The CI check is complaining about `gpg --passphrase $KEY -d -o config/application.yml config/application.yml.asc`. I am not sure how we should deal with this.

File: app/models/reports/all_subscribers.rb [hunk 1]
```diff
@@ -3,17 +3,17 @@ class AllSubscribers < Report
...
     super
   end
 
   def generate(params = {})
-    seasons = params[:seasons]
+    vouchertypes = Report.list_of_ints_from_multiselect(params[:vouchertypes])
     @relation = 
-      if seasons.empty?
+      if vouchertypes.empty? 
       then Customer.purchased_any_vouchertypes(Vouchertype.subscription_vouchertypes.map(&:id))
```
Comment (armandofox) [inline app/models/reports/all_subscribers.rb]: this doesn't make sense.  if no vouchertypes were specified, the report runs anyway?? 
shouldn't the result here be `Customer.none`?

File: features/step_definitions/subscription_steps.rb [hunk 2]
```diff
@@ -4,13 +4,19 @@
...
+
 Given /the following subscribers exist:/ do |subscribers|
   subscribers.hashes.each do |customer|
     first,last = customer['customer'].split(/\s+/)
     cust = find_customer(first,last) || create(:customer, :first_name => first, :last_name => last)
     customer['quantity'] ||= '1'
-    customer['subscriptions'].split(/\s*,\s*/).each do |season|
-      buy!(cust, Vouchertype.find_by(:name => season), customer['quantity'].to_i)
+    customer['subscriptions'].split(/\s*,\s*/).each do |subscription_type|
+      buy!(cust, Vouchertype.find_by(:name => subscription_type), customer['quantity'].to_i)
```
Comment (armandofox) [inline features/step_definitions/subscription_steps.rb]: are you certain this doesn't break other tests that rely on the current step definition??  this step is used by other scenarios, do they all pass locally?

File: lib/tasks/create_staging_data.rake [hunk 3]
```diff
@@ -154,10 +154,26 @@ staging = namespace :staging do
...
       :included_vouchers => sub_vouchers.map(&:id).map(&:to_s).zip(Array.new(sub_vouchers.size) { 1 }).to_h)
+    sub = Vouchertype.create!(
+        :category => :bundle,
+        :subscription => true,
+        :price => 80.00,
+        :name => 'Season Subscription 2',
+        :season => Time.this_season,
+        :offer_public => Vouchertype::ANYONE,
+        :included_vouchers => sub_vouchers.map(&:id).map(&:to_s).zip(Array.new(sub_vouchers.size) { 1 }).to_h)
+    sub = Vouchertype.create!(
```
Comment (armandofox) [inline lib/tasks/create_staging_data.rake]: why do you assign the same variable twice?  if you're not going to use the variable don't assign it.  this is confusing usage.  if all you're doing is creating more sub vouchertypes you should put the creation code in a loop that just changes the name each time - please be DRY, do not copy-and-paste code.

File: lib/tasks/create_staging_data.rake [hunk 4]
```diff
@@ -172,15 +188,24 @@ staging = namespace :staging do
...
+      where(:vouchertypes => {:category => :bundle, :subscription => true, :price => 70.00}).
+      first
+    sub_voucher2 = ValidVoucher.includes(:vouchertype).
+      where(:vouchertypes => {:category => :bundle, :subscription => true, :price => 80.00}).
       first
+    sub_voucher3 = ValidVoucher.includes(:vouchertype).
+      where(:vouchertypes => {:category => :bundle, :subscription => true, :price => 90.00}).
+      first
+    sub_vouchers = [sub_voucher1, sub_voucher2, sub_voucher3]
+
```
Comment (armandofox) [inline lib/tasks/create_staging_data.rake]: same comment applies.  please fix this.  i cannot accept pull requests with copy-pasted code, this is a habit you should have kicked after cs61a!

File: app/models/reports/all_subscribers.rb (hunk 1 above)
Comment (Hexhu) [reply app/models/reports/all_subscribers.rb]: The intention was to preserve the behavior when no seasons were selected, which defaults to a select all, so we left line 15 unchanged. `Customer.none` is indeed more intuitive. Should we make this change?

Comment (armandofox) [reply app/models/reports/all_subscribers.rb]: ok, i see.  it's probably fine to leave the behavior as-is then.  thanks for clarifying.

File: lib/tasks/create_staging_data.rake (hunk 4 above)
Comment (AlanHe-Xiaoyu) [reply lib/tasks/create_staging_data.rake]: We're fixing this right now.

File: features/step_definitions/subscription_steps.rb (hunk 2 above)
Comment (AlanHe-Xiaoyu) [reply features/step_definitions/subscription_steps.rb]: We just double-checked by running all cucumber local tests, and they all passed
224 scenarios (224 passed)
1810 steps (1810 passed)
7m33.514s

File: lib/tasks/create_staging_data.rake (hunk 3 above)
Comment (chengchenghuang) [reply lib/tasks/create_staging_data.rake]: Sorry for the repetition here, I have reverted reate_staging_data.rake back to its original version. I will open another feature branch for adding more subscription_vouchertypes.

File: lib/tasks/create_staging_data.rake (hunk 4 above)
Comment (chengchenghuang) [reply lib/tasks/create_staging_data.rake]: Same here. Sorry for the repetition here, I have reverted reate_staging_data.rake back to its original version. I will open another feature branch for adding more subscription_vouchertypes.

**Comment by Hexhu** [comment]
New proposed commit comment after squashing:

Implement the feature "All subscribers report by subscription voucher types"
  * "All subscribers" report should allow filtering by subscription type, not by season
  * new tests for the new "All subscribers" report

**Comment by Hexhu** [comment]
@armandofox We believe we have finished all the requested changes. Could you review the updated files again and squash if it looks good. CC @AlanHe-Xiaoyu @chengchenghuang @Hexhu

**Pull request merged**